### PR TITLE
Point out that behavior might be switched on 2015 and 2018 too one day

### PR DESCRIPTION
**Pull request by est31**
Reword documentation to make it clear that behaviour can be switched on older editions too, one day in the future. It doesn't *have* to be switched, but I think it's good to have it as an option and re-evaluate it a few months/years down the line when e.g. the crates that showed up in crater were broken by different changes in the language already.

cc #25725, #65819, #66145, #84147 , and https://github.com/rust-lang/rust/issues/84133#issuecomment-818005314

### CODE REVIEW

**Comment by est31** [comment]
cc @nikomatsakis @LukasKalbertodt @cuviper

File: library/std/src/primitive_docs.rs [hunk 1]
```diff
@@ -553,8 +553,10 @@ mod prim_pointer {}
 /// # Editions
 ///
 /// Prior to Rust 1.53, arrays did not implement `IntoIterator` by value, so the method call
-/// `array.into_iter()` auto-referenced into a slice iterator. That behavior is preserved in the
-/// 2015 and 2018 editions of Rust for compatability, ignoring `IntoIterator` by value.
+/// `array.into_iter()` auto-referenced into a slice iterator. Right now, the old behaviour
+/// is preserved in the 2015 and 2018 editions of Rust for compatability, ignoring
```
Comment (CryZe) [inline library/std/src/primitive_docs.rs:557]:
```suggestion
/// is preserved in the 2015 and 2018 editions of Rust for compatibility, ignoring
```

Comment (cuviper) [reply library/std/src/primitive_docs.rs:557]: Oops, I had also misspelled that in <https://github.com/rust-lang/rust/pull/84147#pullrequestreview-641476700>.

**Comment by cuviper** [comment]
That clarification seems good to me.

Somewhat aside, you changed "behavior" to "behaviour" -- I don't know if we have an official position for the project documentation, but the US spelling is far more common throughout the repo.

**Comment by est31** [comment]
> the US spelling is far more common throughout the repo.

Hmmm you're right, indeed it is:

```
$ rg "behaviour" {library,compiler}/ | wc -l ; rg "behavior" {library,compiler}/ | wc -l
42
490
```

I've changed it to US spelling.

Also btw., if you check the [summary comment](https://github.com/rust-lang/rust/pull/65819#issuecomment-681132206) of the August 2020 crater run, it lists the following major regression causing crates: colored, lazy_static, jpeg-decoder, ring. Interestingly, only ring has versions at all that use edition 2018, the other crates in that list all use edition 2015 for their latest versions, so I assume this includes the offending versions too. ring has switched to edition 2018 starting with 0.14.0, which means most of its breakage causing versions are on edition 2018.

File: library/std/src/primitive_docs.rs [hunk 2]
```diff
@@ -553,8 +553,10 @@ mod prim_pointer {}
 /// # Editions
 ///
 /// Prior to Rust 1.53, arrays did not implement `IntoIterator` by value, so the method call
-/// `array.into_iter()` auto-referenced into a slice iterator. That behavior is preserved in the
-/// 2015 and 2018 editions of Rust for compatability, ignoring `IntoIterator` by value.
+/// `array.into_iter()` auto-referenced into a slice iterator. Right now, the old behavior
+/// is preserved in the 2015 and 2018 editions of Rust for compatibility, ignoring
+/// `IntoIterator` by value. In the future, the behavior on the 2015 and 2018 edition
+/// might be made consistent to the behaviour of later editions.
```
Comment (cuviper) [inline library/std/src/primitive_docs.rs:559]: One more:
```suggestion
/// might be made consistent to the behavior of later editions.
```

**Comment by cuviper** [comment]
> Also btw., if you check the [summary comment](https://github.com/rust-lang/rust/pull/65819#issuecomment-681132206) of the August 2020 crater run, it lists the following major regression causing crates: colored, lazy_static, jpeg-decoder, ring. Interestingly, only ring has versions at all that use edition 2018, the other crates in that list all use edition 2015 for their latest versions, so I assume this includes the offending versions too. ring has switched to edition 2018 starting with 0.14.0, which means most of its breakage causing versions are on edition 2018.

Are you suggesting that we might want to migrate one edition without the other? If so, I think the only reason to force this at all would be to get rid of the attribute hack entirely, i.e. all or nothing.

**Comment by est31** [comment]
> Are you suggesting that we might want to migrate one edition without the other?

I don't really know. I mainly shared this because I found it interesting. If there is a difference between 2015 and 2018 then it should only be in timing, for the reason you stated. Right now, I think it's better to perform this change as part of edition 2021 so that people can deliberately opt into the breakage. What one should do with 2015 and 2018 can be decided at a later point in time. I don't really want to turn this thread into a discussion about when to migrate edition 2015 and 2018.... maybe I shouldn't have shared that observation, my fault sorry... The point of this PR to explicitly leave the option open, nothing more nothing less.

**Comment by nikomatsakis** [comment]
@bors r-

I'm not sure about this.

**Comment by nikomatsakis** [comment]
Let me explain:

I'm trying to decide what actionable thing we want people to based on this. I guess we're saying that we don't want them to use `.foo()` methods on older editions, because we might break them, and they should instead code against `IntoIter::into_iter` directly or something?

I think it would be good to spell out the advice.

**Comment by est31** [comment]
@nikomatsakis I've added a paragraph to the document. If I'm reading #84513 correctly it's also the plan to change the `array_into_iter` lint to make such suggestions as well.

**Review by nikomatsakis — APPROVED** [review]

**Review by nikomatsakis — APPROVED** [review]

**Comment by jackh726** [comment]
I think this failed in #84675

@bors r-

**Comment by cuviper** [comment]
Same issue as https://github.com/rust-lang/rust/pull/84147#issuecomment-826254457, which I fixed for my test case in commit f6a90ca168dfdf198b75bcdc0aba01915a5bd6fb.

**Pull request merged**